File: Leetcode/dfs_recursion_backtracking/39_Combination Sum/Solution.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(res, new ArrayList<Integer>(), candidates, target, 0);
        return res;
    }
        private void dfs(List<List<Integer>> res, List<Integer> preCombine,
                         int[] candidates, int preSum, int start){
            if(preSum == 0){
                res.add( new ArrayList<Integer>((preCombine)) );
                return;
            }
            if( preSum < 0 ){
                return;
            }
            for(int i = start;  i< candidates.length; i++){
                preCombine.add(candidates[i]);
                dfs(res, preCombine, candidates, preSum-candidates[i], i);
                preCombine.remove(preCombine.size()-1);
            }
        }
}

File: Leetcode/dfs_recursion_backtracking/39_Combination Sum/Solution.cpp
class Solution {
private:
    void dfs(vector<int>& candidates, int target, vector<vector<int>>& res, vector<int>& temp, int start){
        if(target<0)
            return;
        if(target==0){
            res.push_back(temp);
            return;
        }
        for(int i = start; i<candidates.size(); ++i){
            temp.push_back(candidates[i]);
            dfs(candidates, target-candidates[i], res, temp, i);
            temp.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> temp;
        dfs(candidates, target, res, temp, 0);
        return res;
    }
};
